src/emmy/numerical/ode.cljc

```diff
@@ -2,96 +2,171 @@
+                                                          (<= x x1))
+                                                     (format "ODE interpolation request %g out of range [%g, %g]"
+                                                             x x0 x1))
+                                             (.setInterpolatedTime interpolator x)
+                                             (into [] (.getInterpolatedState interpolator)))}))
+                             (a/<!! step-requests))
+                (throw (InterruptedException. "end of integration"))))))
+         (doto (Thread.
+                (fn []
+                    ;; Wait for the first step request before calling integrate.
+                    ;; Our flow control for the CM3 integrator is through the 
+                    ;; StepHandler callback; when `.integrate` is called, the first
+                    ;; step will be computed, and we want to hold the callback from
+                    ;; returning until the client signals that they are done with the 
+                    ;; (very stateful!) interpolation function, in order to prevent 
+                    ;; the generation of another step's worth of interpolation data
+                    ;; before we are ready. Therefore we enforce the invariant that 
+                    ;; a step request will precede the computation of any step, which
+                    ;; is why we pull from the channel here.
+                  (a/<!! step-requests)
+                  (try
+                    (.integrate gbs ode Double/MAX_VALUE)
+                      ;; InterruptedException is the nominal way to exit an integration
+                    (catch InterruptedException _)
+                      ;; If the integrator throws an exception, send that exception through
+                      ;; the `solution-segments` channel so that it may be handled by the consumer
+                      ;; thread.
+                    (catch Throwable t
+                      (a/>!! solution-segments t)))))
+           (.setDaemon true)
+           (.start))
+         (let [next-segment (fn []
+                              (a/>!! step-requests true)
+                              (a/<!! solution-segments))
+               current-segment (atom (next-segment))]
+           (fn
+             ([]
+              (a/>!! step-requests false))
+             ([x]
+              (when (< x (:x0 @current-segment))
+                (throw (IllegalStateException. "Cannot use interpolation function in backwards direction")))
+              (while (> x (:x1 @current-segment))
+                (let [s (next-segment)]
+                  (reset! current-segment s)))
+              ((:f @current-segment) x)))))
+
+       :cljs
+       (let [jsf' (fn [x ys] (double-array (f' x ys)))
+             solver (o/Solver.
+                     jsf'
+                     dimension
+                     (clj->js {:absoluteTolerance epsilon
```

Comment:
not that it really matters for performance, but if you're writing a literal you can just do `#js {:key "value" ...}` to directly create the js array here, instead of `js->clj`.

Reply:
neat thank you!
 

Reply:
fixed